### ISSUE TITLE
Different behaviour of BCEWithLogitsLoss and BCELoss + Sigmoid

### ISSUE THREAD

**Issue by martinarjovsky** [pytorch/pytorch#1939]
In certain situations BCEWithLogitsLoss returns the wrong answer. This seems to happen in the case where the output has singleton dimensions (such as the typical case when it's the output of an nn.Linear in classification).

Snippet to reproduce
```python
import torch
import torch.nn as nn
from torch.autograd import Variable
import numpy as np

sigmoid = nn.Sigmoid()

t = np.round(np.random.rand(64))
o = np.random.rand(64,1) - 0.5

t = Variable(torch.Tensor(t))
o = Variable(torch.Tensor(o))

print(nn.BCEWithLogitsLoss()(o, t))
print(nn.BCELoss()(sigmoid(o), t)) # Different numbers

o = np.random.rand(64) - 0.5
o = Variable(torch.Tensor(o))

print(nn.BCEWithLogitsLoss()(o, t))
print(nn.BCELoss()(sigmoid(o), t)) # Same numbers
```

Thanks :)

**Comment by alykhantejani** [comment]
Thanks for reporting this @martinarjovsky.

@soumith not sure this is a valid use of the losses - i.e. the output and target have different shapes (one is batched and one is not)?

According to the docstrings of `binary_cross_entropy` and `binary_cross_entropy_with_logits` the `input` and `target` should have the same shape. Shall we just add a check and raise a `ValueError` if they are not? Happy to send a PR for this

**Comment by martinarjovsky** [comment]
@alykhantejani Actually, both of these are batched (64 would be the batch size). You have this behaviour in https://github.com/pytorch/examples/blob/master/imagenet/main.py#L169 as well with multiclass classification (which doesn't have this bug), where output is 256 x 1000 and target is 256. In here, you have 64 x 1 and target is 64.

**Comment by alykhantejani** [comment]
@martinarjovsky yes, you're right.

Perhaps we need to do something like:

**Comment by martinarjovsky** [comment]
That would work :)

**Comment by alykhantejani** [comment]
Cool, I'll send this in a PR and add a test to catch this broadcasting issue in the future.

As @fmassa just pointed out to me, it might be better to raise an error if  `input.numel() !=  target.numel()` and then just do `target = target.view(input.size())`. This would avoid any broadcasting

**Comment by soumith** [comment]
yes, this seems sensible here, to avoid broadcasting.thanks @alykhantejani

**Comment by alykhantejani** [comment]
@soumith I think this issue can now be closed?
